sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala

```diff
@@ -2575,6 +2575,22 @@ class AdaptiveQueryExecSuite
       assert(findTopLevelAggregate(adaptive5).size == 4)
     }
   }
+
+  test("SPARK-38578: Avoid unnecessary sort in FileFormatWriter if user has specified sort") {
+    Seq(
+      ("key", "key, value", false),
+      ("key as x", "key, value", false),
+      ("key", "key", true),
+      ("key, value", "key", true)
+    ).foreach { case (project, sort, required) =>
+      val (origin, adaptive) = runAdaptiveAndVerifyResult(
+        s"""
+           |SELECT $project FROM testdata where key < 0 ORDER BY $sort
```

Comment:
It's really hard to add a test of a table insertion with a sort, since the extra sort is added and executed at FileFormatWriter. So I just test it manally if you do not mind, see PR's description.

Reply:
what do you mean by "if needed"? This is a plain SELECT query and I don't think we can remove the sort (or the table `testdata` has an ordering?)

Reply:
its sort has been replaced by empty local relation during re-optimze.

Reply:
Let's add more code comments to explain this test, otherwise most people won't understand it.

Reply:
added some comments to describe the test